### ISSUE TITLE
Is there a way to check if the Scope was disposed?

### ISSUE THREAD

**Issue by ilya-malakhovsky** [simpleinjector/SimpleInjector#976]
Hi!

I'm using `SimpleInjector` with `Transient` and `Singleton` scopes mostly for my web-application (for historical reasons). Now I want to add some monitoring at the end of each web-request to check if state of objects is correct (basically I'm checking that `DBContext` doesn't contain any unsaved changes)

For this I've created a `Tracker` class :
1. `Tracker` registered as `Scoped` in the container
2.  For each creation of `IUnitOfWork` I've injected resolving of `Tracker` from the container and adding created `IUnitOfWork` to tracked list inside the `Tracker`
3. On `Scope.WhenScopeEnds` `Tracker` checks all registered `IUnitOfWork`s

(I can't change the Lifetime of  `IUnitOfWork` registration)

Factory method for `IUnitOfWork` looks like this if simplified:
``` c#
IUnitOfWork Create()
{
    var unitOfWork = new UnitOfWork(...);

    if (Lifestyle.Scoped.GetCurrentScope(container) != null)
    {
        // Track only if the Scope is valid
        container.GetInstance<Tracker>().Register(unitOfWork);
    }

    return unitOfWork;
}
```

My problem is that `IUnitOfWork` can be resolved at any stage of the application (I can't change this) including when there's no web-request yet (I'm checking this as `Lifestyle.Scoped.GetCurrentScope(container) != null`) and when the request has just ended (checks on `HttpApplication.EndRequest`).

In the latter case I sometimes get:
```[ObjectDisposedException] : `Cannot access a disposed object.  Object name: 'SimpleInjector.Scope' ```

Is there a way to know that the `Scope` was already disposed? In the debug I can see that `GetCurrentScope` returns me a disposed `Scope` (I see that it' was disposed via it's internal properties). I would expect it to be null as technically we are out of scope already.

**Comment by dotnetjunkie** [comment]
Can you supply me with a full stack trace of the exception?

**Comment by dotnetjunkie** [comment]
I struggled with the weird request lifestyle model of those web frameworks years ago, which is one of the many reasons I moved to design that I found easier to reason about, which is the use of an `ICommandHandler<T>` abstraction, which I described [here](https://blogs.cuttingedge.it/steven/p/commands/). This allows me to apply cross-cutting concerns like your tracker logic on the boundaries of executed commands.

Changing your application to a design as described in that article might be quite an undertaking, and might not be feasible for you at the moment, but it might give you some inspiration on ways of looking at this from a different angle.
